In [1]:
import pandas as pd
import numpy as np
import os

### Etapa 1 - Criação do dataframe com dados populacionais por ano dos municípios brasileiros para servir de base de cáclulo das variáveis [QUANTIDADE DE PROCEDIMENTOS POR 10 MIL HAB] E [VALORES DE PROCEDIEMNTOS POR 10 MIL HAB]

In [2]:
# Criando função para concatenar poulação dos municípios por ano, entre 2010 e 2022, a partir dos dados em arquivo csv na pasta ./data/demografico/ para uso em cálculo posterior
def load_population_data(folder_path):
    dfs = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.csv'):
            year = int(filename.split('.')[0])
            df = pd.read_csv(os.path.join(folder_path, filename), sep=';', names=['codigo', 'pop'], dtype={'codigo': str})
            df.set_index('codigo', inplace=True)
            df = df[~df.index.duplicated(keep='first')] # remove possíveis linhas duplicadas
            df.rename(columns={'pop': str(year)}, inplace=True)
            dfs.append(df.iloc[1:, :]) # ignora a primeira linha
    merged_df = pd.concat(dfs, axis=1)
    merged_df.fillna(0, inplace=True) # preenche valores nulos com 0
    return merged_df


In [3]:
# Aplicando a função fornecendo a pasta com dados como parâmetro
# OBS: esta pasta deve conter somente estes arquivos. Não colocar outros arquivos nesta pasta pois a funçãqo retornará erro!
df_pop = load_population_data('E:/OneDrive/Documentos/GitHub/iris/data/demografico')  

In [4]:
# checando o dataframe e o resultado da função de concatenação de arquivos csv
df_pop

2010     2011     2012     2013     2014     2015     2016  \
codigo                                                                   
1100015    24392    24228    24069    25728    25652    25578    25506   
1100023    90353    91570    92747   101269   102860   104401   105896   
1100031     6313     6221     6132     6495     6424     6355     6289   
1100049    78574    78958    79330    85863    86556    87226    87877   
1100056    17029    16939    16852    18041    18013    17986    17959   
...          ...      ...      ...      ...      ...      ...      ...   
5222005    12548    12644    12737    13227    13343    13456    13567   
5222054     7371     7475     7576     7933     8053     8171     8286   
5222203     4735     4846     4954     5246     5371     5495     5615   
5222302     5145     5195     5244     5460     5520     5578     5635   
5300108  2570160  2609997  2648532  2789761  2852372  2914830  2977216   

            2017     2018     2019     2020     2021     2022  
codigo                                                         
1100015    25437    23167    22945    22728    22516    22516  
1100023   107345   106168   107863   109523   111148   111148  
1100031     6224     5438     5312     5188     5067     5067  
1100049    88507    84813    85359    85893    86416    86416  
1100056    17934    16444    16323    16204    16088    16088  
...          ...      ...      ...      ...      ...      ...  
5222005    13675    13746    13863    13977    14088    14088  
5222054     8397     8611     8743     8873     9002     9002  
5222203     5731     6026     6171     6312     6451     6451  
5222302     5690     5758     5821     5882     5941     5941  
5300108  3039444  2974703  3015268  3055149  3094325  3094325  

[5570 rows x 13 columns]

In [5]:
# Checando quantidade de municípios
len(df_pop)

5570

In [6]:
# checando tipo de variáveis resultante da função 
df_pop.dtypes

2010    object
2011    object
2012    object
2013    object
2014    object
2015    object
2016    object
2017    object
2018    object
2019    object
2020    object
2021    object
2022    object
dtype: object

In [7]:
# convertendo dados para tipo integer
df_pop = df_pop.astype(int)

In [8]:
# Checando valores iguais a zero em cada coluna
num_zeros_por_coluna = (df_pop == 0).sum()
print(num_zeros_por_coluna)
del num_zeros_por_coluna # exclui a var temp, após verificação

2010    0
2011    0
2012    0
2013    0
2014    0
2015    0
2016    0
2017    0
2018    0
2019    0
2020    0
2021    0
2022    0
dtype: int64


Nota importante: cinco municípios dos 5.570 existentes atualmente não constavam nas bases de dados do IBGE entre 2010 e 2012, provavelmente foram criados após essa data. Os municípios tem os seguintes códigos: 1504752, 4212650, 4220000, 4314548, 5006275. Os dados inexistenets foram preenchidos com cálculo de população decrescente de 2012, para 2011 e para 2010 com abse na diferença da população anual dos anos subsequentes. Desse modo a matriz não resultará em dados nulos quando for utilizada para cálculos com outras matrizes para obtenção de dados relacionados à população, como é o caso da quantidade de procedimentos por 10 mil habitantes. 


In [9]:
import requests
# obtendo dados dos municípios listados por meio da API do IBGE

def obter_dados_municipios(codigos):
    dados_municipios = []
    for codigo in codigos.split("|"):
        try:
            url = f'https://servicodados.ibge.gov.br/api/v1/localidades/municipios/{codigo}'
            response = requests.get(url)
            if response.status_code == 200:
                dados_municipio = response.json()
                nome_municipio = dados_municipio['nome']
                uf_municipio = dados_municipio['microrregiao']['mesorregiao']['UF']['sigla']
                dados_municipios.append({'Nome': nome_municipio, 'UF': uf_municipio})
        except requests.exceptions.RequestException as e:
            print(f"Erro ao obter dados do município {codigo}: {e}")
    return dados_municipios



In [10]:
# fornece parâmetros com os códigos dos municípios para obtenção do nome e UF
codigos = '1504752|4212650|4220000|4314548|5006275'
municipios = obter_dados_municipios(codigos)
for municipio in municipios:
    print(f"Nome: {municipio['Nome']}, UF: {municipio['UF']}")


Nome: Mojuí dos Campos, UF: PA
Nome: Pescaria Brava, UF: SC
Nome: Balneário Rincão, UF: SC
Nome: Pinto Bandeira, UF: RS
Nome: Paraíso das Águas, UF: MS


3 criados em 2013, sendo que Balneário Rincão e Pinto Bandeira foram emancipados antes de 2010. De toda forma os dados de população para os 5 municípios nos 3 anos foram projetados de maneira a não causar viés nos dados.

In [11]:
# mostando os dados dos municípios e as projeções resultantes para população nos anos de 2010 a 2012. 
df_pop.loc[['1504752', '4212650', '4220000', '4314548', '5006275']]


2010   2011   2012   2013   2014   2015   2016   2017   2018   2019  \
codigo                                                                          
1504752  14905  15014  15123  15232  15341  15446  15548  15646  15982  16084   
4212650   9465   9539   9613   9687   9761   9835   9908   9980  10022  10091   
4220000  11040  11236  11432  11628  11824  12018  12212  12403  12570  12760   
4314548   2324   2443   2562   2681   2800   2824   2847   2868   2968   3003   
5006275   4627   4732   4837   4942   5047   5150   5251   5350   5455   5555   

          2020   2021   2022  
codigo                        
1504752  16184  16282  16282  
4212650  10159  10225  10225  
4220000  12946  13129  13129  
4314548   3036   3068   3068  
5006275   5654   5751   5751

### Etapa 2 - Criação do dataframe com dados dos procedimentos para calcular matrizes de treinamento da rede neural autoencoder mlp e dos algoritmos de ML supervisionado e não supervisionado

In [12]:
df_sia = pd.read_csv('./data/dfpo1/SIA_procedimentos.csv', sep='|')
df_sia

cnes  competencia                      estabelecimento  \
0            329       201105  CAPS ESTACAO CIDADANIA                
1            329       201102  CAPS ESTACAO CIDADANIA                
2            329       201111  CAPS ESTACAO CIDADANIA                
3            329       201209  CAPS ESTACAO CIDADANIA                
4            329       201104  CAPS ESTACAO CIDADANIA                
...          ...          ...                                  ...   
7477277  9993827       202009  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477278  9993827       202110  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477279  9993827       202208  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477280  9993827       202207  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477281  9995471       202011  HOFTALMED                             

                                razao_social  uf  codigo_municipio  \
0        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
1        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
2        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
3        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
4        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
...                                      ...  ..               ...   
7477277  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477278  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477279  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477280  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477281  CRISTALINK CLINICA MEDICA LTDA       SC          420420.0   

                       municipio  co_procedimento  \
0        CABO DE SANTO AGOSTINHO         30108006   
1        CABO DE SANTO AGOSTINHO         30108010   
2        CABO DE SANTO AGOSTINHO         30108010   
3        CABO DE SANTO AGOSTINHO         30108010   
4        CABO DE SANTO AGOSTINHO         30108012   
...                          ...              ...   
7477277                   GUACUI         60480001   
7477278                   GUACUI         60483001   
7477279                   GUACUI         60483001   
7477280                   GUACUI         60483002   
7477281                  CHAPECO         40503013   

                                              procedimento  \
0        ACOMPANHAMENTO INTENSIVO DE PACIENTE EM SAUDE ...   
1        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
2        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
3        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
4        ACOMPANHAMENTO SEMI-INTENSIVO DE PACIENTES EM ...   
...                                                    ...   
7477277  ENOXAPARINA SÓDICA 40 MG/0,4 ML SOLUÇÃO INJETÁ...   
7477278  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 50 MG (...   
7477279  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 50 MG (...   
7477280  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 100 MG ...   
7477281                               VITRECTOMIA ANTERIOR   

         qtde_procedimentos  valor_procedimentos  
0                       166              3004.60  
1                        52               772.20  
2                        30               445.50  
3                        42               623.70  
4                       110              1749.00  
...                     ...                  ...  
7477277                  30                 0.00  
7477278                  56               193.76  
7477279                  56               193.76  
7477280                  60               207.60  
7477281                   1               381.08  

[7477282 rows x 11 columns]

In [13]:
df_sia.rename(columns={'co_procedimento': 'codigo_procedimento'}, inplace=True)

In [14]:
df_sih = pd.read_csv('./data/dfpo1/SIH_procedimentos.csv', sep='|')
df_sih

cnes  competencia                           estabelecimento  \
0              35       201002                    HOSPITAL MENDO SAMPAIO   
1              35       201002                    HOSPITAL MENDO SAMPAIO   
2              35       201003                    HOSPITAL MENDO SAMPAIO   
3              35       201003                    HOSPITAL MENDO SAMPAIO   
4              35       201004                    HOSPITAL MENDO SAMPAIO   
...           ...          ...                                       ...   
28657071  9950931       202211  HOSPITAL MUNICIPAL INTEGRADO SANTO AMARO   
28657072  9950931       202211  HOSPITAL MUNICIPAL INTEGRADO SANTO AMARO   
28657073  9994289       202105    POLICLINICA MUNICIPAL ESTEFANIA FARIAS   
28657074  9997784       202107      HOSPITAL E MATERNIDADE NAILA GONCALO   
28657075  9997784       202205      HOSPITAL E MATERNIDADE NAILA GONCALO   

                                             razao_social  uf  \
0         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
1         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
2         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
3         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
4         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
...                                                   ...  ..   
28657071             PREFEITURA DO MUNICIPIO DE SAO PAULO  SP   
28657072             PREFEITURA DO MUNICIPIO DE SAO PAULO  SP   
28657073                  PREFEITURA MUNICIPAL DE SURUBIM  PE   
28657074                          MUNICIPIO DE BACABAEIRA  MA   
28657075                          MUNICIPIO DE BACABAEIRA  MA   

          codigo_municipio                municipio  codigo_procedimento  \
0                   260290  CABO DE SANTO AGOSTINHO            303080078   
1                   260290  CABO DE SANTO AGOSTINHO            303090316   
2                   260290  CABO DE SANTO AGOSTINHO            303020083   
3                   260290  CABO DE SANTO AGOSTINHO            409060135   
4                   260290  CABO DE SANTO AGOSTINHO            303070102   
...                    ...                      ...                  ...   
28657071            355030                SAO PAULO            407040099   
28657072            355030                SAO PAULO            407040102   
28657073            261450                  SURUBIM            303010223   
28657074            210125                BACABEIRA            411010034   
28657075            210125                BACABEIRA            310010039   

                                               procedimento  \
0                             TRATAMENTO DE ESTREPTOCOCCIAS   
1              TRATAMENTO DAS POLIARTROPATIAS INFLAMATORIAS   
2         TRATAMENTO DE OUTRAS DOENÇAS DO SANGUE E DOS Ó...   
3                                       HISTERECTOMIA TOTAL   
4         TRATAMENTO DE OUTRAS DOENCAS DO APARELHO DIGES...   
...                                                     ...   
28657071                 HERNIOPLASTIA INGUINAL (BILATERAL)   
28657072       HERNIOPLASTIA INGUINAL / CRURAL (UNILATERAL)   
28657073  TRATAMENTO DE INFECÇÃO PELO CORONAVIRUS – COVI...   
28657074                                    PARTO CESARIANO   
28657075                                       PARTO NORMAL   

          qtde_procedimentos  valor_procedimentos  
0                         11               340.62  
1                         11               238.70  
2                         24               388.37  
3                         12               658.03  
4                         15               380.58  
...                      ...                  ...  
28657071                  27              5490.54  
28657072                  51             10845.49  
28657073                  44             33000.00  
28657074                  74              4867.57  
28657075                  68              3731.10  

[28657076 rows x 11 colu

In [15]:
df_sia['fonte'] = 'sia'
df_sih['fonte'] = 'sih'

In [16]:
# Definindo a opção "float_format" para exibir somente 1 dígito após a vírgula
pd.set_option("float_format", "{:.2f}".format)

In [17]:
df_sia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7477282 entries, 0 to 7477281
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     float64
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(2), int64(4), object(6)
memory usage: 684.6+ MB


In [18]:
df_sih.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28657076 entries, 0 to 28657075
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     int64  
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(1), int64(5), object(6)
memory usage: 2.6+ GB


In [19]:
df_procedimentos = pd.concat([df_sia, df_sih])
df_procedimentos.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 36134358 entries, 0 to 28657075
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     float64
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(2), int64(4), object(6)
memory usage: 3.5+ GB


In [20]:
print(df_procedimentos['fonte'].value_counts() / df_procedimentos['fonte'].count() *100)


sih   79.31
sia   20.69
Name: fonte, dtype: float64


In [21]:
# cria um df com base em uma pivot table para os dados de quantidade de procedimento por cnes, com colunas com dados do município
df_qtde_proc_cnes = (
    df_procedimentos
    .pivot_table(index=['codigo_municipio', 'municipio', 'uf', 'cnes', 'codigo_procedimento', 'procedimento'], columns='competencia', values='qtde_procedimentos')
    .reset_index()
    .sort_index()
)

In [22]:
# Convertendo o índice "date" em uma coluna
df_qtde_proc_cnes = df_qtde_proc_cnes.reset_index()

# Removendo o nome da coluna "variable" da primeira linha
df_qtde_proc_cnes = df_qtde_proc_cnes.rename_axis(None, axis=1)
df_qtde_proc_cnes  = df_qtde_proc_cnes.drop('index', axis =1)

In [23]:
# modifica tipo de variável para codigo do município para inteiro
df_qtde_proc_cnes['codigo_municipio'] = df_qtde_proc_cnes['codigo_municipio'].astype(int)
df_qtde_proc_cnes

codigo_municipio                            municipio  uf     cnes  \
0                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
1                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
2                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
3                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
4                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
...                   ...                                  ...  ..      ...   
1377489            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377490            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377491            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377492            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377493            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   

         codigo_procedimento  \
0                  301060010   
1                  301060088   
2                  303010010   
3                  303010037   
4                  303010045   
...                      ...   
1377489             70101020   
1377490             70101021   
1377491             70101022   
1377492             70101024   
1377493             70101025   

                                              procedimento  201001  201002  \
0        DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
1        DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
2                            TRATAMENTO DE DENGUE CLÁSSICA  139.00  113.00   
3                 TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
4             TRATAMENTO DE DOENÇAS BACTERIANAS ZOONÓTICAS     NaN     NaN   
...                                                    ...     ...     ...   
1377489                         CADEIRA DE RODAS MONOBLOCO     NaN     NaN   
1377490                      CADEIRA DE RODAS (ACIMA 90KG)     NaN     NaN   
1377491     CADEIRA DE RODAS MOTORIZADA ADULTO OU INFANTIL     NaN     NaN   
1377492  CADEIRA DE RODAS PARA BANHO COM ENCOSTO RECLIN...     NaN     NaN   
1377493   CADEIRA DE RODAS PARA BANHO COM ARO DE PROPULSÃO     NaN     NaN   

         201003  201004  ...  202203  202204  202205  202206  202207  202208  \
0           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2         30.00   19.00  ...   12.00    6.00   28.00  138.00  261.00  189.00   
3         12.00    6.00  ...    4.00   22.00   12.00   56.00  157.00     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
1377489     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377490     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377491     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377492     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377493     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

         202209  202210  202211  202212  
0           NaN   38.00   55.00     NaN  
1           NaN   82.00   94.00     NaN  
2         91.00  134.00  183.00     NaN  
3         63.00  249.00    2.00     NaN  
4           NaN     NaN     NaN     NaN  
...         ...     ...     ...     ...  
1377489     NaN     NaN     NaN     NaN  
1377490     NaN     NaN     NaN     NaN  
1377491     NaN     NaN     NaN     NaN  
1377492     NaN     NaN     NaN     NaN  
1377493     NaN     NaN     NaN     NaN  

[1377494 rows x 162 columns]

In [24]:
df_qtde_proc_cnes.to_csv('./data/dfpo1/df_qtde_proc_cnes.csv', sep = '|', encoding = 'UTF-8')

In [25]:
# verifica a média de valores NAN mas colunas...  que se mostra muito alto, pois se temos um df com 1,37 milhões de linhas e a média de linhas com valores NaN é de 1,1 milhões, 
# então precisamos selecionar somente os procedimentos com poucos valores ausentes nas linhas tanto para CNES quanto apra municípios
df_qtde_proc_cnes.isna().sum().mean()

1103507.2654320989

In [26]:
df_qtde_proc_mun = (
    df_procedimentos
    .pivot_table(index=['codigo_municipio', 'municipio', 'uf', 'codigo_procedimento', 'procedimento'], columns='competencia', values='qtde_procedimentos')
    .reset_index()
    .sort_index()
)

In [30]:
# Convertendo o índice "date" em uma coluna
df_qtde_proc_mun = df_qtde_proc_mun.reset_index()

# Removendo o nome da coluna "variable" da primeira linha
df_qtde_proc_mun = df_qtde_proc_mun.rename_axis(None, axis=1)
df_qtde_proc_mun  = df_qtde_proc_mun.drop('index', axis =1)
df_qtde_proc_mun

codigo_municipio                            municipio  uf  \
0              110001.00                ALTA FLORESTA D'OESTE  RO   
1              110001.00                ALTA FLORESTA D'OESTE  RO   
2              110001.00                ALTA FLORESTA D'OESTE  RO   
3              110001.00                ALTA FLORESTA D'OESTE  RO   
4              110001.00                ALTA FLORESTA D'OESTE  RO   
...                  ...                                  ...  ..   
932634         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932635         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932636         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932637         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932638         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   

        codigo_procedimento  \
0                  30108012   
1                 301060010   
2                 301060088   
3                 303010010   
4                 303010037   
...                     ...   
932634            416120032   
932635            416120040   
932636            416120059   
932637            505010097   
932638            505010119   

                                             procedimento  201001  201002  \
0       ACOMPANHAMENTO SEMI-INTENSIVO DE PACIENTES EM ...     NaN  168.00   
1       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
2       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
3                           TRATAMENTO DE DENGUE CLÁSSICA  139.00  113.00   
4                TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
...                                                   ...     ...     ...   
932634                   MASTECTOMIA SIMPLES EM ONCOLOGIA     NaN     NaN   
932635  RESSECÇÃO DE LESÃO NÃO PALPÁVEL DE MAMA COM MA...     NaN     NaN   
932636  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...     NaN     NaN   
932637                              TRANSPLANTE DE CORNEA   54.00   48.00   
932638             TRANSPLANTE DE CORNEA (EM REOPERACOES)     NaN     NaN   

        201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0       168.00  768.00  840.00  ...     NaN     NaN     NaN     NaN     NaN   
1          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
2          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
3        30.00   19.00    4.00  ...   12.00    6.00   28.00  138.00  261.00   
4        12.00    6.00   10.00  ...    4.00   22.00   12.00   56.00  157.00   
...        ...     ...     ...  ...     ...     ...     ...     ...     ...   
932634   47.00    5.00   27.00  ...     NaN     NaN     NaN     NaN     NaN   
932635     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932636     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932637   72.00   33.00   18.00  ...     NaN     NaN     NaN     NaN     NaN   
932638     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   

        202208  202209  202210  202211  202212  
0          NaN     NaN     NaN     NaN     NaN  
1          NaN     NaN   38.00   55.00     NaN  
2          NaN     NaN   82.00   94.00     NaN  
3       189.00   91.00  134.00  183.00     NaN  
4          NaN   63.00  249.00    2.00     NaN  
...        ...     ...     ...     ...     ...  
932634     NaN     NaN     NaN     NaN     NaN  
932635     NaN     NaN     NaN     NaN     NaN  
932636     NaN     NaN     NaN     NaN     NaN  
932637     NaN     NaN     NaN     NaN     NaN  
932638     NaN     NaN     NaN     NaN     NaN  

[932639 rows x 161 columns]

In [28]:
df_qtde_proc_mun.to_csv('./data/dfpo1/df_qtde_proc_mun.csv', sep = '|', encoding = 'UTF-8')

In [32]:
df_qtde_proc_mun

codigo_municipio                            municipio  uf  \
0              110001.00                ALTA FLORESTA D'OESTE  RO   
1              110001.00                ALTA FLORESTA D'OESTE  RO   
2              110001.00                ALTA FLORESTA D'OESTE  RO   
3              110001.00                ALTA FLORESTA D'OESTE  RO   
4              110001.00                ALTA FLORESTA D'OESTE  RO   
...                  ...                                  ...  ..   
932634         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932635         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932636         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932637         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932638         530180.00  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   

        codigo_procedimento  \
0                  30108012   
1                 301060010   
2                 301060088   
3                 303010010   
4                 303010037   
...                     ...   
932634            416120032   
932635            416120040   
932636            416120059   
932637            505010097   
932638            505010119   

                                             procedimento  201001  201002  \
0       ACOMPANHAMENTO SEMI-INTENSIVO DE PACIENTES EM ...     NaN  168.00   
1       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
2       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
3                           TRATAMENTO DE DENGUE CLÁSSICA  139.00  113.00   
4                TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
...                                                   ...     ...     ...   
932634                   MASTECTOMIA SIMPLES EM ONCOLOGIA     NaN     NaN   
932635  RESSECÇÃO DE LESÃO NÃO PALPÁVEL DE MAMA COM MA...     NaN     NaN   
932636  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...     NaN     NaN   
932637                              TRANSPLANTE DE CORNEA   54.00   48.00   
932638             TRANSPLANTE DE CORNEA (EM REOPERACOES)     NaN     NaN   

        201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0       168.00  768.00  840.00  ...     NaN     NaN     NaN     NaN     NaN   
1          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
2          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
3        30.00   19.00    4.00  ...   12.00    6.00   28.00  138.00  261.00   
4        12.00    6.00   10.00  ...    4.00   22.00   12.00   56.00  157.00   
...        ...     ...     ...  ...     ...     ...     ...     ...     ...   
932634   47.00    5.00   27.00  ...     NaN     NaN     NaN     NaN     NaN   
932635     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932636     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932637   72.00   33.00   18.00  ...     NaN     NaN     NaN     NaN     NaN   
932638     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   

        202208  202209  202210  202211  202212  
0          NaN     NaN     NaN     NaN     NaN  
1          NaN     NaN   38.00   55.00     NaN  
2          NaN     NaN   82.00   94.00     NaN  
3       189.00   91.00  134.00  183.00     NaN  
4          NaN   63.00  249.00    2.00     NaN  
...        ...     ...     ...     ...     ...  
932634     NaN     NaN     NaN     NaN     NaN  
932635     NaN     NaN     NaN     NaN     NaN  
932636     NaN     NaN     NaN     NaN     NaN  
932637     NaN     NaN     NaN     NaN     NaN  
932638     NaN     NaN     NaN     NaN     NaN  

[932639 rows x 161 columns]

In [33]:
df_qtde_proc_mun.isna().sum().mean()

728710.354037267

In [34]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_cnes.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
procedimentos_cnes_sem_nan = df_qtde_proc_cnes.dropna(thresh=min_valid_values).reset_index()

# Mantém apenas as colunas 'codigo_procedimento' e 'procedimento' e remove duplicatas
procedimentos_cnes_sem_nan = procedimentos_cnes_sem_nan[['codigo_procedimento', 'procedimento']].drop_duplicates()


In [35]:
procedimentos_cnes_sem_nan

codigo_procedimento                                       procedimento
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...
...                    ...                                                ...
54561            408060638  TRATAMENTO CIRÚRGICO DE LUXAÇÃO / FRATURA-LUXA...
54844            303080086                       TRATAMENTO DE FARMACODERMIAS
55480            408030321  ARTRODESE TORACO-LOMBO-SACRA POSTERIOR, SETE N...
55482            408050667  TRATAMENTO CIRÚRGICO DE LESÃO AGUDA CAPSULO-LI...
55483            408060018            ALONGAMENTO / ENCURTAMENTO MIOTENDINOSO

[921 rows x 2 columns]

In [36]:
procedimentos_cnes_validos = len(procedimentos_cnes_sem_nan)

In [37]:
procedimentos_cnes_sem_nan.to_excel('./data/dfpo1/procedimentos_cnes_iris.xlsx')

NOTA 2: isso implica que no total, seriam 921 os procedimentos POR CNES com no mínimo 90% de valores válidos (não nulos) nas linhas que poderiam ser analizados pela IRIS sem que seja necesário artifícios de tratamento e substituição de missings por dados sintéticos que causem viéses nos dados (substituição de missings por dados calculados, como média ou mediana. Sendo que, para a solução proposta conforme método de auditoria do MS, a mediana seria mais adequada para não repetir os viéses dos outliers de possíveis fraudes e porque é a medida utilizada nos processos de auditoria como no caso do Maranhão)

In [38]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_mun.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
procedimentos_mun_sem_nan = df_qtde_proc_mun.dropna(thresh=min_valid_values).reset_index()

# Mantém apenas as colunas 'codigo_procedimento' e 'procedimento' e remove duplicatas
procedimentos_mun_sem_nan = procedimentos_mun_sem_nan[['codigo_procedimento', 'procedimento']].drop_duplicates()

procedimentos_mun_sem_nan

codigo_procedimento                                       procedimento
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...
...                    ...                                                ...
48522            409060267                 SALPINGOPLASTIA VIDEOLAPAROSCOPICA
48542            412030098                                       PLEUROSTOMIA
48733            408060409               RETIRADA DE TRAÇÃO TRANS-ESQUELÉTICA
52951            408030321  ARTRODESE TORACO-LOMBO-SACRA POSTERIOR, SETE N...
53018            413040097                                 PREPARO DE RETALHO

[1176 rows x 2 columns]

In [39]:
procedimentos_mun_validos = len(procedimentos_mun_sem_nan)
procedimentos_mun_validos

1176

In [40]:
procedimentos_mun_sem_nan.to_excel('./data/dfpo1/procedimentos_mun_iris.xlsx')

### df_proc_cnes é o dataframe com as quantidades de procedimentos POR CNES com menos de 10% de valores NaN nas linhas (ou seja, nos 13 anos ou 156 colunas de competencia mes/ano)
### Este df deve ser a base da qual calculamos todas as demais (df com as distâncias das medianas e assim por diante.. e os mesmos procedimentos devem ser feitos com os df com valores)

In [41]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_cnes.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
df_proc_cnes = df_qtde_proc_cnes.dropna(thresh=min_valid_values).reset_index()

df_proc_cnes = df_proc_cnes.drop('index', axis =1)

df_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento                                       procedimento  \
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS   
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA   
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   
...                    ...                                                ...   
55600             70103013  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55601             70103014  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55602             50301001  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   
55603             50401002                  PROCESSAMENTO DE CORNEA / ESCLERA   
55604             30501016  MANUTENÇÃO E ACOMPANHAMENTO DOMICILIAR DE PACI...   

       201001  201002  201003  201004  ...  202203  202204  202205  202206  \
0        4.00    9.00   12.00    6.00  ...    4.00   22.00   12.00   56.00   
1      116.00  109.00  137.00   89.00  ...   16.00   23.00   14.00   70.00   
2       12.00     NaN    4.00    7.00  ...   20.00   16.00    8.00   76.00   
3        9.00   11.00   37.00    4.00  ...   14.00   16.00   16.00   11.00   
4       10.00   18.00   10.00   12.00  ...   10.00    8.00   10.00    3.00   
...       ...     ...     ...     ...  ...     ...     ...     ...     ...   
55600   16.00   84.00   65.00   72.00  ...   63.00   63.00   31.00   63.00   
55601   14.00   36.00   92.00   30.00  ...   27.00   27.00   27.00   27.00   
55602     NaN   62.00   77.00   80.00  ...   62.00   82.00   70.00   99.00   
55603     NaN  129.00  170.00  198.00  ...  158.00  155.00  165.00  227.00   
55604     NaN     NaN     NaN     NaN  ... 2027.00 3045.00 1912.00 1949.00   

       202207  202208  202209  202210  202211  202212  
0      157.00     NaN   63.00  249.00    2.00     NaN  
1      105.00  108.00    6.00   45.00  163.00     NaN  
2      307.00   87.00  120.00   46.00  220.00     NaN  
3       91.00   27.00   50.00    8.00   37.00     NaN  
4       33.00  258.00    4.00   17.00   12.00     NaN  
...       ...     ...     ...     ...     ...     ...  
55600   63.00     NaN     NaN     NaN   13.00   38.00  
55601   27.00     NaN     NaN     NaN    6.00   19.00  
55602   53.00   62.00   29.00   66.00   40.00     NaN  
55603  119.00  142.00   42.00  150.00   88.00     NaN  
55604  864.00 2061.00 2071.00 3545.00 2028.00 1385.00  

[55605 rows x 162 columns]

### df_proc_mun é o dataframe com as quantidades de procedimentos POR MUNICÍPIO com menos de 10% de valores NaN nas linhas (ou seja, nos 13 anos ou 156 colunas de competencia mes/ano)
### Este df deve ser a base da qual calculamos todas as demais (df com as distâncias das medianas, df com a quantidade de procedimentos para cada 10 mil hab. e assim por diante.. e os mesmos procedimentos devem ser feitos com os df com valores)

In [42]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_mun.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
df_proc_mun = df_qtde_proc_mun.dropna(thresh=min_valid_values).reset_index()

# drop coluna desnecassária criada nos procedimetnos anteriores
df_proc_mun = df_proc_mun.drop('index', axis =1)

# modifica tipo de variável para codigo do município para inteiro
df_proc_mun['codigo_municipio'] = df_proc_mun['codigo_municipio'].astype(int)

df_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS  116.00  109.00   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   12.00     NaN   
3                       TRATAMENTO DE CRISE HIPERTENSIVA    9.00   11.00   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   10.00   18.00   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...   87.50   29.00   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...    6.00     NaN   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   16.00  184.00   
53031                              TRANSPLANTE DE CORNEA   31.00    9.00   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...  955.00  560.50   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0       12.00    6.00   10.00  ...    4.00   22.00   12.00   56.00  157.00   
1      137.00   89.00   74.00  ...   16.00   23.00   14.00   70.00  105.00   
2        4.00    7.00    9.00  ...   20.00   16.00    8.00   76.00  307.00   
3       37.00    4.00   19.00  ...   14.00   16.00   16.00   11.00   91.00   
4       10.00   12.00    8.00  ...   10.00    8.00   10.00    3.00   33.00   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028   16.00   46.33   49.50  ...   34.00   43.25   19.80   17.80   27.50   
53029   56.50    4.00   23.00  ...   76.60   34.00   55.50   40.40   29.83   
53030   13.00   26.00   57.00  ...   78.00   69.00   27.25  117.67   53.00   
53031   47.50   20.50     NaN  ...   67.50  116.00   59.50   65.50   55.00   
53032  986.00 1023.50  570.50  ... 1664.75 3103.33 2100.75 2082.67 2336.50   

       202208  202209  202210  202211  202212  
0         NaN   63.00  249.00    2.00     NaN  
1      108.00    6.00   45.00  163.00     NaN  
2       87.00  120.00   46.00  220.00     NaN  
3       27.00   50.00    8.00   37.00     NaN  
4      258.00    4.00   17.00   12.00     NaN  
...       ...     ...     ...     ...     ...  
53028   16.00   27.80  123.33   35.00   47.80  
53029   64.60   48.80  123.00   59.50   47.67  
53030  149.50   70.67   52.00   81.00   26.67  
53031   29.00   39.00   84.50  145.50    5.00  
53032 1707.25  630.67  915.25 1520.67  478.00  

[53033 rows x 161 columns]

In [43]:
# Selecionar apenas colunas numéricas após a quarta coluna
num_cols = df_proc_mun.columns[5:]

# Substituindo NaN pela mediana de cada linha
df_proc_mun[num_cols] = df_proc_mun[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)

# Calculando o desvio da mediana para cada célula
desvios_med_n_mun = df_proc_mun[num_cols].apply(lambda x: x - x.median(), axis=1)

# Criando um novo DataFrame com os desvios_med_n como uma matriz (m)
m_desvios_med_n_mun = pd.DataFrame(desvios_med_n_mun.values, index=df_proc_mun.index, columns=num_cols)

# Concatenando as quatro primeiras colunas do DataFrame original com o novo DataFrame de desvios_med_n_mun
df_desv_med_n_proc_mun = pd.concat([df_proc_mun.iloc[:, :5], m_desvios_med_n_mun], axis=1)
df_desv_med_n_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS   -4.00    1.00   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   79.00   72.00   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   -5.00    0.00   
3                       TRATAMENTO DE CRISE HIPERTENSIVA  -11.00   -9.00   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   -4.00    4.00   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...   37.50  -21.00   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...  -29.00    0.00   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...  -86.83   81.17   
53031                              TRANSPLANTE DE CORNEA  -20.00  -42.00   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D... -169.83 -564.33   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0        4.00   -2.00    2.00  ...   -4.00   14.00    4.00   48.00  149.00   
1      100.00   52.00   37.00  ...  -21.00  -14.00  -23.00   33.00   68.00   
2      -13.00  -10.00   -8.00  ...    3.00   -1.00   -9.00   59.00  290.00   
3       17.00  -16.00   -1.00  ...   -6.00   -4.00   -4.00   -9.00   71.00   
4       -4.00   -2.00   -6.00  ...   -4.00   -6.00   -4.00  -11.00   19.00   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028  -34.00   -3.67   -0.50  ...  -16.00   -6.75  -30.20  -32.20  -22.50   
53029   21.50  -31.00  -12.00  ...   41.60   -1.00   20.50    5.40   -5.17   
53030  -89.83  -76.83  -45.83  ...  -24.83  -33.83  -75.58   14.83  -49.83   
53031   -3.50  -30.50    0.00  ...   16.50   65.00    8.50   14.50    4.00   
53032 -138.83 -101.33 -554.33  ...  539.92 1978.50  975.92  957.83 1211.67   

       202208  202209  202210  202211  202212  
0        0.00   55.00  241.00   -6.00    0.00  
1       71.00  -31.00    8.00  126.00    0.00  
2       70.00  103.00   29.00  203.00    0.00  
3        7.00   30.00  -12.00   17.00    0.00  
4      244.00  -10.00    3.00   -2.00    0.00  
...       ...     ...     ...     ...     ...  
53028  -34.00  -22.20   73.33  -15.00   -2.20  
53029   29.60   13.80   88.00   24.50   12.67  
53030   46.67  -32.17  -50.83  -21.83  -76.17  
53031  -22.00  -12.00   33.50   94.50  -46.00  
53032  582.42 -494.17 -209.58  395.83 -646.83  

[53033 rows x 161 columns]

In [44]:
df_desv_med_n_proc_mun.to_excel('./data/dfpo1/df_desv_med_n_proc_mun.xlsx')

In [45]:
df_proc_mun[num_cols] = df_proc_mun[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)
df_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS  116.00  109.00   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   12.00   17.00   
3                       TRATAMENTO DE CRISE HIPERTENSIVA    9.00   11.00   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   10.00   18.00   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...   87.50   29.00   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...    6.00   35.00   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   16.00  184.00   
53031                              TRANSPLANTE DE CORNEA   31.00    9.00   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...  955.00  560.50   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0       12.00    6.00   10.00  ...    4.00   22.00   12.00   56.00  157.00   
1      137.00   89.00   74.00  ...   16.00   23.00   14.00   70.00  105.00   
2        4.00    7.00    9.00  ...   20.00   16.00    8.00   76.00  307.00   
3       37.00    4.00   19.00  ...   14.00   16.00   16.00   11.00   91.00   
4       10.00   12.00    8.00  ...   10.00    8.00   10.00    3.00   33.00   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028   16.00   46.33   49.50  ...   34.00   43.25   19.80   17.80   27.50   
53029   56.50    4.00   23.00  ...   76.60   34.00   55.50   40.40   29.83   
53030   13.00   26.00   57.00  ...   78.00   69.00   27.25  117.67   53.00   
53031   47.50   20.50   51.00  ...   67.50  116.00   59.50   65.50   55.00   
53032  986.00 1023.50  570.50  ... 1664.75 3103.33 2100.75 2082.67 2336.50   

       202208  202209  202210  202211  202212  
0        8.00   63.00  249.00    2.00    8.00  
1      108.00    6.00   45.00  163.00   37.00  
2       87.00  120.00   46.00  220.00   17.00  
3       27.00   50.00    8.00   37.00   20.00  
4      258.00    4.00   17.00   12.00   14.00  
...       ...     ...     ...     ...     ...  
53028   16.00   27.80  123.33   35.00   47.80  
53029   64.60   48.80  123.00   59.50   47.67  
53030  149.50   70.67   52.00   81.00   26.67  
53031   29.00   39.00   84.50  145.50    5.00  
53032 1707.25  630.67  915.25 1520.67  478.00  

[53033 rows x 161 columns]

In [46]:
df_proc_mun.to_excel('./data/dfpo1/df_proc_mun.xlsx')

In [47]:
# Selecionar apenas colunas numéricas após a quarta coluna
num_cols = df_proc_cnes.columns[6:]

# Substituindo NaN pela mediana de cada linha
df_proc_cnes[num_cols] = df_proc_cnes[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)

# Calculando o desvio da mediana para cada célula
desvios_med_n_cnes = df_proc_cnes[num_cols].apply(lambda x: x - x.median(), axis=1)

# Criando um novo DataFrame com os desvios_med_n como uma matriz (m)
m_desvios_med_n_cnes = pd.DataFrame(desvios_med_n_cnes.values, index=df_proc_cnes.index, columns=num_cols)

# Concatenando as quatro primeiras colunas do DataFrame original com o novo DataFrame de desvios_med_n_cnes
df_desv_med_n_proc_cnes = pd.concat([df_proc_cnes.iloc[:, :5], m_desvios_med_n_cnes], axis=1)
df_desv_med_n_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento  201001  201002  201003  201004  201005  ...  \
0                303010037   -4.00    1.00    4.00   -2.00    2.00  ...   
1                303010061   79.00   72.00  100.00   52.00   37.00  ...   
2                303010193   -5.00    0.00  -13.00  -10.00   -8.00  ...   
3                303060107  -11.00   -9.00   17.00  -16.00   -1.00  ...   
4                303070064   -4.00    4.00   -4.00   -2.00   -6.00  ...   
...                    ...     ...     ...     ...     ...     ...  ...   
55600             70103013  -47.00   21.00    2.00    9.00   28.00  ...   
55601             70103014  -13.00    9.00   65.00    3.00   15.00  ...   
55602             50301001    0.00  -15.00    0.00    3.00    0.00  ...   
55603             50401002    0.00  -25.00   16.00   44.00   36.00  ...   
55604             30501016    0.00    0.00    0.00    0.00    0.00  ...   

       202203  202204  202205  202206  202207  202208  202209  202210  202211  \
0       -4.00   14.00    4.00   48.00  149.00    0.00   55.00  241.00   -6.00   
1      -21.00  -14.00  -23.00   33.00   68.00   71.00  -31.00    8.00  126.00   
2        3.00   -1.00   -9.00   59.00  290.00   70.00  103.00   29.00  203.00   
3       -6.00   -4.00   -4.00   -9.00   71.00    7.00   30.00  -12.00   17.00   
4       -4.00   -6.00   -4.00  -11.00   19.00  244.00  -10.00    3.00   -2.00   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
55600    0.00    0.00  -32.00    0.00    0.00    0.00    0.00    0.00  -50.00   
55601    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  -21.00   
55602  -15.00    5.00   -7.00   22.00  -24.00  -15.00  -48.00  -11.00  -37.00   
55603    4.00    1.00   11.00   73.00  -35.00  -12.00 -112.00   -4.00  -66.00   
55604  440.00 1458.00  325.00  362.00 -723.00  474.00  484.00 1958.00  441.00   

       202212  
0        0.00  
1        0.00  
2        0.00  
3        0.00  
4        0.00  
...       ...  
55600  -25.00  
55601   -8.00  
55602    0.00  
55603    0.00  
55604 -202.00  

[55605 rows x 161 columns]

In [48]:
df_desv_med_n_proc_cnes.to_excel('./data/dfpo1/df_desv_med_n_proc_cnes.xlsx')

In [49]:
df_proc_cnes[num_cols] = df_proc_cnes[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)
df_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento                                       procedimento  \
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS   
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA   
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   
...                    ...                                                ...   
55600             70103013  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55601             70103014  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55602             50301001  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   
55603             50401002                  PROCESSAMENTO DE CORNEA / ESCLERA   
55604             30501016  MANUTENÇÃO E ACOMPANHAMENTO DOMICILIAR DE PACI...   

       201001  201002  201003  201004  ...  202203  202204  202205  202206  \
0        4.00    9.00   12.00    6.00  ...    4.00   22.00   12.00   56.00   
1      116.00  109.00  137.00   89.00  ...   16.00   23.00   14.00   70.00   
2       12.00   17.00    4.00    7.00  ...   20.00   16.00    8.00   76.00   
3        9.00   11.00   37.00    4.00  ...   14.00   16.00   16.00   11.00   
4       10.00   18.00   10.00   12.00  ...   10.00    8.00   10.00    3.00   
...       ...     ...     ...     ...  ...     ...     ...     ...     ...   
55600   16.00   84.00   65.00   72.00  ...   63.00   63.00   31.00   63.00   
55601   14.00   36.00   92.00   30.00  ...   27.00   27.00   27.00   27.00   
55602   77.00   62.00   77.00   80.00  ...   62.00   82.00   70.00   99.00   
55603  154.00  129.00  170.00  198.00  ...  158.00  155.00  165.00  227.00   
55604 1587.00 1587.00 1587.00 1587.00  ... 2027.00 3045.00 1912.00 1949.00   

       202207  202208  202209  202210  202211  202212  
0      157.00    8.00   63.00  249.00    2.00    8.00  
1      105.00  108.00    6.00   45.00  163.00   37.00  
2      307.00   87.00  120.00   46.00  220.00   17.00  
3       91.00   27.00   50.00    8.00   37.00   20.00  
4       33.00  258.00    4.00   17.00   12.00   14.00  
...       ...     ...     ...     ...     ...     ...  
55600   63.00   63.00   63.00   63.00   13.00   38.00  
55601   27.00   27.00   27.00   27.00    6.00   19.00  
55602   53.00   62.00   29.00   66.00   40.00   77.00  
55603  119.00  142.00   42.00  150.00   88.00  154.00  
55604  864.00 2061.00 2071.00 3545.00 2028.00 1385.00  

[55605 rows x 162 columns]

In [50]:
df_proc_cnes.to_excel('./data/dfpo1/df_proc_cnes.xlsx')

### Calculando o df_proc_pop com os dados de quantidade de procedimentos para cada 10 mil habitantes para os 13 anos. 

In [51]:
df_proc_mun.to_excel('./data/dfpo1/df_proc_mun.xlsx')

In [52]:
df_pop

2010     2011     2012     2013     2014     2015     2016  \
codigo                                                                   
1100015    24392    24228    24069    25728    25652    25578    25506   
1100023    90353    91570    92747   101269   102860   104401   105896   
1100031     6313     6221     6132     6495     6424     6355     6289   
1100049    78574    78958    79330    85863    86556    87226    87877   
1100056    17029    16939    16852    18041    18013    17986    17959   
...          ...      ...      ...      ...      ...      ...      ...   
5222005    12548    12644    12737    13227    13343    13456    13567   
5222054     7371     7475     7576     7933     8053     8171     8286   
5222203     4735     4846     4954     5246     5371     5495     5615   
5222302     5145     5195     5244     5460     5520     5578     5635   
5300108  2570160  2609997  2648532  2789761  2852372  2914830  2977216   

            2017     2018     2019     2020     2021     2022  
codigo                                                         
1100015    25437    23167    22945    22728    22516    22516  
1100023   107345   106168   107863   109523   111148   111148  
1100031     6224     5438     5312     5188     5067     5067  
1100049    88507    84813    85359    85893    86416    86416  
1100056    17934    16444    16323    16204    16088    16088  
...          ...      ...      ...      ...      ...      ...  
5222005    13675    13746    13863    13977    14088    14088  
5222054     8397     8611     8743     8873     9002     9002  
5222203     5731     6026     6171     6312     6451     6451  
5222302     5690     5758     5821     5882     5941     5941  
5300108  3039444  2974703  3015268  3055149  3094325  3094325  

[5570 rows x 13 columns]

Os dataframes serão salvos em formato pickle para que estejam disponíveis em outros notebooks

In [53]:
# salvar os dataframes em um arquivo pickle

### POR CNES
# df procedimentos por cnes
df_proc_cnes.to_pickle('./data/dfpo1/df_proc_cnes.pkl')
# matriz de desvios médios por cnes
m_desvios_med_n_cnes.to_pickle('./data/dfpo1/m_desvios_med_n_cnes.pkl')

### POR MUNICÍPIO
# df procedimentos por município
df_proc_mun.to_pickle('./data/dfpo1/df_proc_mun.pkl')
# df populações dos municípios
df_pop.to_pickle('./data/dfpo1/df_pop.pkl')
# matriz de desvios médios por cnes
m_desvios_med_n_mun.to_pickle('./data/dfpo1/m_desvios_med_n_mun.pkl')

In [120]:
# Resetar o índice do df_pop para tornar "codigo" uma coluna
df_pop_reset = df_pop.reset_index()

# Agora, podemos seguir o mesmo processo de antes
df_pop_melt = df_pop_reset.melt(id_vars='codigo', var_name='ano', value_name='pop')

# Repetir cada valor de população 12 vezes (para cada mês)
df_pop_melt = df_pop_melt.loc[df_pop_melt.index.repeat(12)].reset_index(drop=True)

# Criar uma nova coluna para o mês e preenchê-la com valores de 1 a 12 repetidos para cada ano
df_pop_melt['mes'] = np.tile(np.arange(1, 13), len(df_pop_melt) // 12)

# Criar uma nova coluna 'anomes' combinando o ano e o mês
df_pop_melt['anomes'] = df_pop_melt['ano'].astype(str) + df_pop_melt['mes'].apply(lambda x: f'{x:02d}')

# Ajustar o df_pop_melt para ter o mesmo formato que df_proc_municipio
df_pop_reshaped = df_pop_melt.pivot(index='codigo', columns='anomes', values='pop')

# Converter o índice e as colunas de volta para inteiros
df_pop_reshaped.index = df_pop_reshaped.index.astype(int)
df_pop_reshaped.columns = df_pop_reshaped.columns.astype(int)

In [121]:
df_pop_reshaped

anomes    201001   201002   201003   201004   201005   201006   201007  \
codigo                                                                   
1100015    24392    24392    24392    24392    24392    24392    24392   
1100023    90353    90353    90353    90353    90353    90353    90353   
1100031     6313     6313     6313     6313     6313     6313     6313   
1100049    78574    78574    78574    78574    78574    78574    78574   
1100056    17029    17029    17029    17029    17029    17029    17029   
...          ...      ...      ...      ...      ...      ...      ...   
5222005    12548    12548    12548    12548    12548    12548    12548   
5222054     7371     7371     7371     7371     7371     7371     7371   
5222203     4735     4735     4735     4735     4735     4735     4735   
5222302     5145     5145     5145     5145     5145     5145     5145   
5300108  2570160  2570160  2570160  2570160  2570160  2570160  2570160   

anomes    201008   201009   201010  ...   202203   202204   202205   202206  \
codigo                              ...                                       
1100015    24392    24392    24392  ...    22516    22516    22516    22516   
1100023    90353    90353    90353  ...   111148   111148   111148   111148   
1100031     6313     6313     6313  ...     5067     5067     5067     5067   
1100049    78574    78574    78574  ...    86416    86416    86416    86416   
1100056    17029    17029    17029  ...    16088    16088    16088    16088   
...          ...      ...      ...  ...      ...      ...      ...      ...   
5222005    12548    12548    12548  ...    14088    14088    14088    14088   
5222054     7371     7371     7371  ...     9002     9002     9002     9002   
5222203     4735     4735     4735  ...     6451     6451     6451     6451   
5222302     5145     5145     5145  ...     5941     5941     5941     5941   
5300108  2570160  2570160  2570160  ...  3094325  3094325  3094325  3094325   

anomes    202207   202208   202209   202210   202211   202212  
codigo                                                         
1100015    22516    22516    22516    22516    22516    22516  
1100023   111148   111148   111148   111148   111148   111148  
1100031     5067     5067     5067     5067     5067     5067  
1100049    86416    86416    86416    86416    86416    86416  
1100056    16088    16088    16088    16088    16088    16088  
...          ...      ...      ...      ...      ...      ...  
5222005    14088    14088    14088    14088    14088    14088  
5222054     9002     9002     9002     9002     9002     9002  
5222203     6451     6451     6451     6451     6451     6451  
5222302     5941     5941     5941     5941     5941     5941  
5300108  3094325  3094325  3094325  3094325  3094325  3094325  

[5570 rows x 156 columns]

In [122]:
# Se 'codigo' é um nível de índice e não uma coluna, você pode usar reset_index para torná-lo uma coluna novamente.
df_pop_reshaped.reset_index(level=0, inplace=True)

# Como resultado, agora você deve ter 'codigo' como uma coluna normal.


In [123]:
df_pop_reshaped

anomes   codigo   201001   201002   201003   201004   201005   201006  \
0       1100015    24392    24392    24392    24392    24392    24392   
1       1100023    90353    90353    90353    90353    90353    90353   
2       1100031     6313     6313     6313     6313     6313     6313   
3       1100049    78574    78574    78574    78574    78574    78574   
4       1100056    17029    17029    17029    17029    17029    17029   
...         ...      ...      ...      ...      ...      ...      ...   
5565    5222005    12548    12548    12548    12548    12548    12548   
5566    5222054     7371     7371     7371     7371     7371     7371   
5567    5222203     4735     4735     4735     4735     4735     4735   
5568    5222302     5145     5145     5145     5145     5145     5145   
5569    5300108  2570160  2570160  2570160  2570160  2570160  2570160   

anomes   201007   201008   201009  ...   202203   202204   202205   202206  \
0         24392    24392    24392  ...    22516    22516    22516    22516   
1         90353    90353    90353  ...   111148   111148   111148   111148   
2          6313     6313     6313  ...     5067     5067     5067     5067   
3         78574    78574    78574  ...    86416    86416    86416    86416   
4         17029    17029    17029  ...    16088    16088    16088    16088   
...         ...      ...      ...  ...      ...      ...      ...      ...   
5565      12548    12548    12548  ...    14088    14088    14088    14088   
5566       7371     7371     7371  ...     9002     9002     9002     9002   
5567       4735     4735     4735  ...     6451     6451     6451     6451   
5568       5145     5145     5145  ...     5941     5941     5941     5941   
5569    2570160  2570160  2570160  ...  3094325  3094325  3094325  3094325   

anomes   202207   202208   202209   202210   202211   202212  
0         22516    22516    22516    22516    22516    22516  
1        111148   111148   111148   111148   111148   111148  
2          5067     5067     5067     5067     5067     5067  
3         86416    86416    86416    86416    86416    86416  
4         16088    16088    16088    16088    16088    16088  
...         ...      ...      ...      ...      ...      ...  
5565      14088    14088    14088    14088    14088    14088  
5566       9002     9002     9002     9002     9002     9002  
5567       6451     6451     6451     6451     6451     6451  
5568       5941     5941     5941     5941     5941     5941  
5569    3094325  3094325  3094325  3094325  3094325  3094325  

[5570 rows x 157 columns]

In [ ]:
df_proc_mun = df_proc_mun.drop('index', axis=1)

In [133]:
df_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    4.00    9.00   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS  116.00  109.00   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   12.00   17.00   
3                       TRATAMENTO DE CRISE HIPERTENSIVA    9.00   11.00   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   10.00   18.00   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...   87.50   29.00   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...    6.00   35.00   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   16.00  184.00   
53031                              TRANSPLANTE DE CORNEA   31.00    9.00   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...  955.00  560.50   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0       12.00    6.00   10.00  ...    4.00   22.00   12.00   56.00  157.00   
1      137.00   89.00   74.00  ...   16.00   23.00   14.00   70.00  105.00   
2        4.00    7.00    9.00  ...   20.00   16.00    8.00   76.00  307.00   
3       37.00    4.00   19.00  ...   14.00   16.00   16.00   11.00   91.00   
4       10.00   12.00    8.00  ...   10.00    8.00   10.00    3.00   33.00   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028   16.00   46.33   49.50  ...   34.00   43.25   19.80   17.80   27.50   
53029   56.50    4.00   23.00  ...   76.60   34.00   55.50   40.40   29.83   
53030   13.00   26.00   57.00  ...   78.00   69.00   27.25  117.67   53.00   
53031   47.50   20.50   51.00  ...   67.50  116.00   59.50   65.50   55.00   
53032  986.00 1023.50  570.50  ... 1664.75 3103.33 2100.75 2082.67 2336.50   

       202208  202209  202210  202211  202212  
0        8.00   63.00  249.00    2.00    8.00  
1      108.00    6.00   45.00  163.00   37.00  
2       87.00  120.00   46.00  220.00   17.00  
3       27.00   50.00    8.00   37.00   20.00  
4      258.00    4.00   17.00   12.00   14.00  
...       ...     ...     ...     ...     ...  
53028   16.00   27.80  123.33   35.00   47.80  
53029   64.60   48.80  123.00   59.50   47.67  
53030  149.50   70.67   52.00   81.00   26.67  
53031   29.00   39.00   84.50  145.50    5.00  
53032 1707.25  630.67  915.25 1520.67  478.00  

[53033 rows x 161 columns]

In [134]:
df_proc_mun.dtypes

codigo_municipio         int64
municipio               object
uf                      object
codigo_procedimento      int64
procedimento            object
                        ...   
202208                 float64
202209                 float64
202210                 float64
202211                 float64
202212                 float64
Length: 161, dtype: object

In [136]:
df_pop_reshaped.dtypes

anomes
codigo    int64
201001    int32
201002    int32
201003    int32
201004    int32
          ...  
202208    int32
202209    int32
202210    int32
202211    int32
202212    int32
Length: 157, dtype: object

In [137]:
df_pop_reshaped

anomes   codigo   201001   201002   201003   201004   201005   201006  \
0       1100015    24392    24392    24392    24392    24392    24392   
1       1100023    90353    90353    90353    90353    90353    90353   
2       1100031     6313     6313     6313     6313     6313     6313   
3       1100049    78574    78574    78574    78574    78574    78574   
4       1100056    17029    17029    17029    17029    17029    17029   
...         ...      ...      ...      ...      ...      ...      ...   
5565    5222005    12548    12548    12548    12548    12548    12548   
5566    5222054     7371     7371     7371     7371     7371     7371   
5567    5222203     4735     4735     4735     4735     4735     4735   
5568    5222302     5145     5145     5145     5145     5145     5145   
5569    5300108  2570160  2570160  2570160  2570160  2570160  2570160   

anomes   201007   201008   201009  ...   202203   202204   202205   202206  \
0         24392    24392    24392  ...    22516    22516    22516    22516   
1         90353    90353    90353  ...   111148   111148   111148   111148   
2          6313     6313     6313  ...     5067     5067     5067     5067   
3         78574    78574    78574  ...    86416    86416    86416    86416   
4         17029    17029    17029  ...    16088    16088    16088    16088   
...         ...      ...      ...  ...      ...      ...      ...      ...   
5565      12548    12548    12548  ...    14088    14088    14088    14088   
5566       7371     7371     7371  ...     9002     9002     9002     9002   
5567       4735     4735     4735  ...     6451     6451     6451     6451   
5568       5145     5145     5145  ...     5941     5941     5941     5941   
5569    2570160  2570160  2570160  ...  3094325  3094325  3094325  3094325   

anomes   202207   202208   202209   202210   202211   202212  
0         22516    22516    22516    22516    22516    22516  
1        111148   111148   111148   111148   111148   111148  
2          5067     5067     5067     5067     5067     5067  
3         86416    86416    86416    86416    86416    86416  
4         16088    16088    16088    16088    16088    16088  
...         ...      ...      ...      ...      ...      ...  
5565      14088    14088    14088    14088    14088    14088  
5566       9002     9002     9002     9002     9002     9002  
5567       6451     6451     6451     6451     6451     6451  
5568       5941     5941     5941     5941     5941     5941  
5569    3094325  3094325  3094325  3094325  3094325  3094325  

[5570 rows x 157 columns]

In [157]:
df_proc_mun.codigo_municipio.unique()

array([110001, 110002, 110003, ..., 522185, 522200, 530010], dtype=int64)

In [160]:
df_pop_reshaped

anomes   codigo   201001   201002   201003   201004   201005   201006  \
0       1100015    24392    24392    24392    24392    24392    24392   
1       1100023    90353    90353    90353    90353    90353    90353   
2       1100031     6313     6313     6313     6313     6313     6313   
3       1100049    78574    78574    78574    78574    78574    78574   
4       1100056    17029    17029    17029    17029    17029    17029   
...         ...      ...      ...      ...      ...      ...      ...   
5565    5222005    12548    12548    12548    12548    12548    12548   
5566    5222054     7371     7371     7371     7371     7371     7371   
5567    5222203     4735     4735     4735     4735     4735     4735   
5568    5222302     5145     5145     5145     5145     5145     5145   
5569    5300108  2570160  2570160  2570160  2570160  2570160  2570160   

anomes   201007   201008   201009  ...   202203   202204   202205   202206  \
0         24392    24392    24392  ...    22516    22516    22516    22516   
1         90353    90353    90353  ...   111148   111148   111148   111148   
2          6313     6313     6313  ...     5067     5067     5067     5067   
3         78574    78574    78574  ...    86416    86416    86416    86416   
4         17029    17029    17029  ...    16088    16088    16088    16088   
...         ...      ...      ...  ...      ...      ...      ...      ...   
5565      12548    12548    12548  ...    14088    14088    14088    14088   
5566       7371     7371     7371  ...     9002     9002     9002     9002   
5567       4735     4735     4735  ...     6451     6451     6451     6451   
5568       5145     5145     5145  ...     5941     5941     5941     5941   
5569    2570160  2570160  2570160  ...  3094325  3094325  3094325  3094325   

anomes   202207   202208   202209   202210   202211   202212  
0         22516    22516    22516    22516    22516    22516  
1        111148   111148   111148   111148   111148   111148  
2          5067     5067     5067     5067     5067     5067  
3         86416    86416    86416    86416    86416    86416  
4         16088    16088    16088    16088    16088    16088  
...         ...      ...      ...      ...      ...      ...  
5565      14088    14088    14088    14088    14088    14088  
5566       9002     9002     9002     9002     9002     9002  
5567       6451     6451     6451     6451     6451     6451  
5568       5941     5941     5941     5941     5941     5941  
5569    3094325  3094325  3094325  3094325  3094325  3094325  

[5570 rows x 157 columns]

In [163]:
df_pop_reshaped.dtypes

anomes
codigo    int64
201001    int32
201002    int32
201003    int32
201004    int32
          ...  
202208    int32
202209    int32
202210    int32
202211    int32
202212    int32
Length: 157, dtype: object

In [166]:
df_pop_reshaped['codigo'].astype(str).str[:-1].astype(int)

0       110001
1       110002
2       110003
3       110004
4       110005
         ...  
5565    522200
5566    522205
5567    522220
5568    522230
5569    530010
Name: codigo, Length: 5570, dtype: int32

In [169]:
# Criar uma cópia de df_pop_reshaped e adicionar uma nova coluna com o código de município ajustado
df_pop_reshaped['codigo_municipio'] = df_pop_reshaped['codigo'].astype(str).str[:-1].astype(int)

In [170]:
df_pop_reshaped

anomes   codigo   201001   201002   201003   201004   201005   201006  \
0       1100015    24392    24392    24392    24392    24392    24392   
1       1100023    90353    90353    90353    90353    90353    90353   
2       1100031     6313     6313     6313     6313     6313     6313   
3       1100049    78574    78574    78574    78574    78574    78574   
4       1100056    17029    17029    17029    17029    17029    17029   
...         ...      ...      ...      ...      ...      ...      ...   
5565    5222005    12548    12548    12548    12548    12548    12548   
5566    5222054     7371     7371     7371     7371     7371     7371   
5567    5222203     4735     4735     4735     4735     4735     4735   
5568    5222302     5145     5145     5145     5145     5145     5145   
5569    5300108  2570160  2570160  2570160  2570160  2570160  2570160   

anomes   201007   201008   201009  ...   202204   202205   202206   202207  \
0         24392    24392    24392  ...    22516    22516    22516    22516   
1         90353    90353    90353  ...   111148   111148   111148   111148   
2          6313     6313     6313  ...     5067     5067     5067     5067   
3         78574    78574    78574  ...    86416    86416    86416    86416   
4         17029    17029    17029  ...    16088    16088    16088    16088   
...         ...      ...      ...  ...      ...      ...      ...      ...   
5565      12548    12548    12548  ...    14088    14088    14088    14088   
5566       7371     7371     7371  ...     9002     9002     9002     9002   
5567       4735     4735     4735  ...     6451     6451     6451     6451   
5568       5145     5145     5145  ...     5941     5941     5941     5941   
5569    2570160  2570160  2570160  ...  3094325  3094325  3094325  3094325   

anomes   202208   202209   202210   202211   202212  codigo_municipio  
0         22516    22516    22516    22516    22516            110001  
1        111148   111148   111148   111148   111148            110002  
2          5067     5067     5067     5067     5067            110003  
3         86416    86416    86416    86416    86416            110004  
4         16088    16088    16088    16088    16088            110005  
...         ...      ...      ...      ...      ...               ...  
5565      14088    14088    14088    14088    14088            522200  
5566       9002     9002     9002     9002     9002            522205  
5567       6451     6451     6451     6451     6451            522220  
5568       5941     5941     5941     5941     5941            522230  
5569    3094325  3094325  3094325  3094325  3094325            530010  

[5570 rows x 158 columns]

In [174]:
df_pop_reshaped

anomes             codigo   201001   201002   201003   201004   201005  \
codigo_municipio                                                         
110001            1100015    24392    24392    24392    24392    24392   
110002            1100023    90353    90353    90353    90353    90353   
110003            1100031     6313     6313     6313     6313     6313   
110004            1100049    78574    78574    78574    78574    78574   
110005            1100056    17029    17029    17029    17029    17029   
...                   ...      ...      ...      ...      ...      ...   
522200            5222005    12548    12548    12548    12548    12548   
522205            5222054     7371     7371     7371     7371     7371   
522220            5222203     4735     4735     4735     4735     4735   
522230            5222302     5145     5145     5145     5145     5145   
530010            5300108  2570160  2570160  2570160  2570160  2570160   

anomes             201006   201007   201008   201009  ...   202203   202204  \
codigo_municipio                                      ...                     
110001              24392    24392    24392    24392  ...    22516    22516   
110002              90353    90353    90353    90353  ...   111148   111148   
110003               6313     6313     6313     6313  ...     5067     5067   
110004              78574    78574    78574    78574  ...    86416    86416   
110005              17029    17029    17029    17029  ...    16088    16088   
...                   ...      ...      ...      ...  ...      ...      ...   
522200              12548    12548    12548    12548  ...    14088    14088   
522205               7371     7371     7371     7371  ...     9002     9002   
522220               4735     4735     4735     4735  ...     6451     6451   
522230               5145     5145     5145     5145  ...     5941     5941   
530010            2570160  2570160  2570160  2570160  ...  3094325  3094325   

anomes             202205   202206   202207   202208   202209   202210  \
codigo_municipio                                                         
110001              22516    22516    22516    22516    22516    22516   
110002             111148   111148   111148   111148   111148   111148   
110003               5067     5067     5067     5067     5067     5067   
110004              86416    86416    86416    86416    86416    86416   
110005              16088    16088    16088    16088    16088    16088   
...                   ...      ...      ...      ...      ...      ...   
522200              14088    14088    14088    14088    14088    14088   
522205               9002     9002     9002     9002     9002     9002   
522220               6451     6451     6451     6451     6451     6451   
522230               5941     5941     5941     5941     5941     5941   
530010            3094325  3094325  3094325  3094325  3094325  3094325   

anomes             202211   202212  
codigo_municipio                    
110001              22516    22516  
110002             111148   111148  
110003               5067     5067  
110004              86416    86416  
110005              16088    16088  
...                   ...      ...  
522200              14088    14088  
522205               9002     9002  
522220               6451     6451  
522230               5941     5941  
530010            3094325  3094325  

[5570 rows x 157 columns]

In [178]:
df_pop_reshaped.reset_index(level=0, inplace=True)

In [179]:
df_pop_reshaped

anomes  codigo_municipio   codigo   201001   201002   201003   201004  \
0                 110001  1100015    24392    24392    24392    24392   
1                 110002  1100023    90353    90353    90353    90353   
2                 110003  1100031     6313     6313     6313     6313   
3                 110004  1100049    78574    78574    78574    78574   
4                 110005  1100056    17029    17029    17029    17029   
...                  ...      ...      ...      ...      ...      ...   
5565              522200  5222005    12548    12548    12548    12548   
5566              522205  5222054     7371     7371     7371     7371   
5567              522220  5222203     4735     4735     4735     4735   
5568              522230  5222302     5145     5145     5145     5145   
5569              530010  5300108  2570160  2570160  2570160  2570160   

anomes   201005   201006   201007   201008  ...   202203   202204   202205  \
0         24392    24392    24392    24392  ...    22516    22516    22516   
1         90353    90353    90353    90353  ...   111148   111148   111148   
2          6313     6313     6313     6313  ...     5067     5067     5067   
3         78574    78574    78574    78574  ...    86416    86416    86416   
4         17029    17029    17029    17029  ...    16088    16088    16088   
...         ...      ...      ...      ...  ...      ...      ...      ...   
5565      12548    12548    12548    12548  ...    14088    14088    14088   
5566       7371     7371     7371     7371  ...     9002     9002     9002   
5567       4735     4735     4735     4735  ...     6451     6451     6451   
5568       5145     5145     5145     5145  ...     5941     5941     5941   
5569    2570160  2570160  2570160  2570160  ...  3094325  3094325  3094325   

anomes   202206   202207   202208   202209   202210   202211   202212  
0         22516    22516    22516    22516    22516    22516    22516  
1        111148   111148   111148   111148   111148   111148   111148  
2          5067     5067     5067     5067     5067     5067     5067  
3         86416    86416    86416    86416    86416    86416    86416  
4         16088    16088    16088    16088    16088    16088    16088  
...         ...      ...      ...      ...      ...      ...      ...  
5565      14088    14088    14088    14088    14088    14088    14088  
5566       9002     9002     9002     9002     9002     9002     9002  
5567       6451     6451     6451     6451     6451     6451     6451  
5568       5941     5941     5941     5941     5941     5941     5941  
5569    3094325  3094325  3094325  3094325  3094325  3094325  3094325  

[5570 rows x 158 columns]

In [182]:
# Criar uma cópia de df_pop_reshaped e adicionar uma nova coluna com o código de município ajustado
df_pop_reshaped['codigo_municipio'] = df_pop_reshaped['codigo'].astype(str).str[:-1].astype(int)

# Definir a nova coluna "codigo_municipio" como índice do DataFrame
df_pop_reshaped.set_index('codigo_municipio', inplace=True)

# Redefinir o índice do DataFrame df_proc_mun
df_proc_mun.reset_index(drop=True, inplace=True)

# Criar uma cópia de df_proc_mun
df_proc_mun_por_10k = df_proc_mun.copy()

# Selecionar as colunas de população em df_pop_reshaped
pop_columns = df_pop_reshaped.columns[1:]

# Mapear as linhas e colunas corretamente
for i in range(len(df_proc_mun)):
    codigo_municipio = df_proc_mun.loc[i, 'codigo_municipio']
    df_proc_mun_por_10k.iloc[i, 5:] = (df_proc_mun.iloc[i, 5:] / df_pop_reshaped.loc[codigo_municipio, pop_columns]) * 10000

# Verificar se há valores NaN após o cálculo
nan_count = df_proc_mun_por_10k.isna().sum().sum()
print("Quantidade de NaN após o cálculo:", nan_count)


Quantidade de NaN após o cálculo: 0


In [183]:
pop_columns

Index([201001, 201002, 201003, 201004, 201005, 201006, 201007, 201008, 201009,
       201010,
       ...
       202203, 202204, 202205, 202206, 202207, 202208, 202209, 202210, 202211,
       202212],
      dtype='object', name='anomes', length=156)

In [184]:
df_proc_mun_por_10k

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    1.64    3.69   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   47.56   44.69   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...    4.92    6.97   
3                       TRATAMENTO DE CRISE HIPERTENSIVA    3.69    4.51   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...    4.10    7.38   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...    0.34    0.11   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...    0.02    0.14   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...    0.06    0.72   
53031                              TRANSPLANTE DE CORNEA    0.12    0.04   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...    3.72    2.18   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0        4.92    2.46    4.10  ...    1.78    9.77    5.33   24.87   69.73   
1       56.17   36.49   30.34  ...    7.11   10.21    6.22   31.09   46.63   
2        1.64    2.87    3.69  ...    8.88    7.11    3.55   33.75  136.35   
3       15.17    1.64    7.79  ...    6.22    7.11    7.11    4.89   40.42   
4        4.10    4.92    3.28  ...    4.44    3.55    4.44    1.33   14.66   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028    0.06    0.18    0.19  ...    0.11    0.14    0.06    0.06    0.09   
53029    0.22    0.02    0.09  ...    0.25    0.11    0.18    0.13    0.10   
53030    0.05    0.10    0.22  ...    0.25    0.22    0.09    0.38    0.17   
53031    0.18    0.08    0.20  ...    0.22    0.37    0.19    0.21    0.18   
53032    3.84    3.98    2.22  ...    5.38   10.03    6.79    6.73    7.55   

       202208  202209  202210  202211  202212  
0        3.55   27.98  110.59    0.89    3.55  
1       47.97    2.66   19.99   72.39   16.43  
2       38.64   53.30   20.43   97.71    7.55  
3       11.99   22.21    3.55   16.43    8.88  
4      114.59    1.78    7.55    5.33    6.22  
...       ...     ...     ...     ...     ...  
53028    0.05    0.09    0.40    0.11    0.15  
53029    0.21    0.16    0.40    0.19    0.15  
53030    0.48    0.23    0.17    0.26    0.09  
53031    0.09    0.13    0.27    0.47    0.02  
53032    5.52    2.04    2.96    4.91    1.54  

[53033 rows x 161 columns]

In [185]:
# Verificar quantidade de NaN por procedimento por município em df_proc_mun_por_10k
nan_counts = df_proc_mun_por_10k.iloc[:, 5:].isna().sum()

# Verificar quantidade de NaN em df_proc_mun
nan_counts_original = df_proc_mun.iloc[:, 5:].isna().sum()

# Comparar os resultados
comparison = pd.concat([nan_counts, nan_counts_original], axis=1)
comparison.columns = ['NaN em df_proc_mun_por_10k', 'NaN em df_proc_mun']
print(comparison)


        NaN em df_proc_mun_por_10k  NaN em df_proc_mun
201001                           0                   0
201002                           0                   0
201003                           0                   0
201004                           0                   0
201005                           0                   0
...                            ...                 ...
202208                           0                   0
202209                           0                   0
202210                           0                   0
202211                           0                   0
202212                           0                   0

[156 rows x 2 columns]


In [ ]:
### POR MUNICÍPIO
# df quantidade de procedimentos para cada 10 mil habintantes do município
df_proc_mun_por_10k.to_pickle('./data/dfpo1/df_proc_mun_por_10k.pkl')